feat: First-class caching (#1104)
DAGWorks-Inc/hamilton: docs/concepts/caching.rst

Using the Hamilton ``@tag`` decorator, you can cache a node's result to a specific file format (``JSON``, ``CSV``, ``Parquet``, etc.) and circumvent the limitations of ``pickle`` files.

.. code-block:: python

    # my_dataflow.py
    import pandas as pd
    from hamilton.function_modifiers import tag

    def raw_data(path: str) -> pd.DataFrame:
        return pd.read_csv(path)

    @tag(cache="parquet")
    def clean_dataset(raw_data: pd.DataFrame) -> pd.DataFrame:
        raw_data = raw_data.fillna(0)
        return raw_data

    @tag(cache="json")
    def statistics(clean_dataset: pd.DataFrame) -> dict:
        return ...

.. note::

    This aims to match and replace the features of ``hamilton.experimental.h_cache.CachingGraphAdater`` but the API is preliminary.

Resume from checkpoint
-----------------------

The caching feature stores results after each node execution. If execution fails half-way, the ``Driver`` can use the cache metadata and results to resume the run. You can think of it as automatically passing ``.execute(overrides={...})`` from disk.

> To resume, you can specify ``.with_cache(resume_from=...)`` which accepts a specific ``run_id`` or the string ``"latest"`` which will automatically get the results from the latest execution, whether it succeeded or failed. The ``run_id`` is automatically generated internally and can be viewed via the cache logger.

Reviewer: This ignores fingerprinting right? 
E.g.
1. run dataflow with `latest=True`. 
2. there is nothing in the cache -- so compute everything.
3. it fails due to a code error
4. we fix the code.
5. we restart with `latest=True`.
6. since there is data in the cache, for any node names that are the same, we will load the prior value, and continue computation from where it left off.

Author: It doesn't ignore fingerprinting and I believe doesn't work as you expect. 

The `resume_from=` options means "resume from this specific `run_id`". If you have previous runs ("1", "2", "3", "4"), calling `resume_from="latest"` will pick the latest available `run_id` at Driver build time (i.e., `"4"`).

Node names are never used for caching. If code versions changed, this will be a cache miss. However, if your DAG has nodes
`foo -> bar -> baz`. 
1. You execute it
2. `baz` fails
3. you tweak baz
4. you rebuild the Driver with `resume_from="latest"`
5. execute again
6. execution will start at `bar` assuming its code version didn't change

.. code-block:: python

    from hamilton import driver
    import my_dataflow

    dr = (
        driver.Builder()
        .with_modules(my_dataflow)
        .with_cache(resume_from="latest")
        .build()
    )

    dr.execute(...)

.. note::

    Using ``resume_from="latest"`` gets the latest run at the moment of building the ``Driver``. Consequently, calling ``Driver.execute()`` multiple times afterwards will always retrieve the same results from that specific point in time.

> 

Reviewer: There needs to be a  "Only recompute what has changed" section to complement the above "checkpointing section".

E.g. 
2. run dataflow with caching.
3. there is nothing in the cache -- so compute everything.
4. it runs through to completion
5. we then run the code again but change an input
6. we only recompute the parts of the dataflow impacted by that new data
7. we then run the code again, but change code
8. we only recompute the parts of the dataflow impacted by that code change

Author: What you describe as "only recompute what has changed" is the default behavior described under "Basic". Unless you mean something else

Control caching behavior
-------------------------

Hamilton's default caching behavior is ideal to get started and typically what you want during iterative development. However, you might want more control over the cache behavior in specific scenario or when moving to production. The caching behavior can be specified at the node-level via either:

- **Dataflow definition** by adding to the function modifiers ``@tag(caching_behavior="ignore")``
- **Driver definition** by passing a list of node name to ``.with_cache(ignore=["foo", "bar"])``

> The "Driver definition" approach will always override the behavior specified using ``@tag``. Passing empty lists to ``ignore=[]`` effectively disables all ``@tag``. Querying node names directly from the ``Driver`` is an effective way to select multiple nodes at once:

Reviewer: we probably want `None` here versus empty list.

Reviewer: or instead be explicit:
```python
.with_cache(ignore="*")
```

Author: I believe what I meant here is the opposite of what you describe. The section refers to a dataflow like this

```python
@tag(cache_behavior="ignore")
def  foo() -> int: ...

def bar(foo: int) -> int: ...
```

What if the user wants to "disable" `cache="ignore"` from the `Driver` code? To me, having
```python
dr = driver.Builder().with_modules(...).with_cache(ignore=[ ])
```
Means "disregard all `cache_behavior="ignore"` statements". Since there's more than 2 behaviors, "disregard all `ignore`" isn't equivalent to "set all to `default`". 

```python
.with_cache(ignore="*")
#not equal to 
.with_cache(default="*")
```

    def current_temperature(weather_client: weather_api.Client) -> float:
        """Get the current local temperature from a public API"""
        return weather_api.Client(**credentials).get_temperature(...)

    def weather_message(
        current_temperature: float, current_wind: dict, is_rain: bool,
    ) -> str:
        """Interpret the temperature and return a message"""
        if ...:
            return "what a nice day 🌞"
        elif ...:
            return "you should bring your umbrella ☔"
        else:
            return "it's starting to feel cold 🥶"

Don't fingerprint
~~~~~~~~~~~~~~~~~~

.. list-table::
   :header-rows: 1
   :widths: auto
   :align: center

   * - Try to retrieve result
     - Store result
     - Fingerprint result
     - Store fingerprint
   * - ✅
     - ✅
     - ❌
     - ❌

Large and complex data objects (e.g., dataframes, machine learning models) can be expensive or unreliable to fingerprint, but may be valuable to cache nonetheless.

In that case, using ``caching_behavior="dont_fingerprint"`` will store the result with a constant fingerprint making the result only retrievable via ``resume_from="latest"``.

> For instance, we don't want to store the ``weather_message`` because it's only a set of ``if/else`` that are trivial to compute.

Reviewer: where is weather message?

        return weather_api.Client(**credentials)

    def current_temperature(weather_client: weather_api.Client) -> float:
        """Get the current local temperature from a public API"""
        return weather_api.Client(**credentials).get_temperature(...)

    @tag(caching_behavior="dont_fingerprint")
    def weather_predictor(
        current_temperature: float, current_wind: dict, is_rain: bool,
    ) -> str:
        """Train a model to predict the probability of rain"""
        model = HistGradientBoostingRegressor(...)
        model.train(...)
        return model

.. note::

     An alternative solution is to register a custom fingerprinting function by ``type`` (see **Fingerprinting**).

Don't store result
~~~~~~~~~~~~~~~~~~~

.. list-table::
   :header-rows: 1
   :widths: auto
   :align: center

   * - Try to retrieve result
     - Store result
     - Fingerprint result
     - Store fingerprint
   * - ✅
     - ❌
     - ✅
     - ❌

If a node produces results that are large in size, not picklable, or very inexpensive to compute, it might be preferrable to not store it. Also, your environment may have storage or I/O limitations (e.g., web service).

> Specifying ``caching_behavior="dont_store_result"`` means Hamilton won't add the result to the result store, but the fingerprint won't be added to the metadata store since it would point to nothing (i.e., "result store miss"). The fingerprint is only used to create the key for children nodes (e.g., ``recommended_outfit``).

Reviewer: ```suggestion
Specifying ``caching_behavior="dont_store_result"`` means Hamilton won't add the result to the result store, and the fingerprint won't be added to the metadata store since it would point to nothing (i.e., "result store miss"). The fingerprint is only used to create the key for downstream nodes (e.g., ``recommended_outfit``).
```

        return weather_api.Client(**credentials).get_temperature(...)

    @tag(caching_behavior="dont_store_result")
    def weather_message(
        current_temperature: float,
        current_wind: dict,
        is_rain: bool,
    ) -> str:
        """Interpret the temperature and return a message"""
        if ...:
            return "what a nice day 🌞"
        elif ...:
            return "you should bring your umbrella ☔"
        else:
            return "it's starting to feel cold 🥶"

    def recommend_outfit(ll_client: llm.Client, weather_message: str) -> str:
        """Use an LLM to generate an outfit suggestion based on the weather"""
        prompt = f"Make an outfit suggestion based on: {weather_message}""
        response = llm.Client.generate(prompt)
        return response

.. note::

    Even when ``caching_behavior="dont_store_result"``, the adapter will try to retrieve stored values that could have been produce before this behavior was set.

Summary
~~~~~~~~

.. list-table::
   :header-rows: 1
   :widths: auto
   :align: center

   * - Behavior
     - Try to retrieve result
     - Store result
     - Fingerprint result

>      - Store fingerprint

Reviewer: When reading this table I don't know what I need to do.
This should be `@tag(...)` since that's what you want people to do, right?

Author: This is the summary table from other subheaders

Reviewer: I understand -- but I need to scroll up to make that connection.
